app/celery/nightly_tasks.py

```diff
@@ -334,3 +336,57 @@ def save_daily_notification_processing_time(bst_date=None):
             messages_within_10_secs=result.messages_within_10_secs,
         )
     )
+
+
+@notify_celery.task(name="delete-oldest-quarter-of-unneeded-notification-history")
+def delete_oldest_quarter_of_unneeded_notification_history():
+    # This function will delete either 1 or 0 quarters of notifications from the notification_history
+    # table. It will pick the oldest quarter available before our retention limit and delete that.
+    # If there are no notifications older than our retention limit, then it won't do anything.
+    # By running this task several nights in a row, it will slowly delete a quarter of notification_history
+    # per night until we reach out retention limit
+    #
+    # We do everything in this function in BST up until the final point
+    # This is because we care about quarters in the BST timezone because this is
+    # how we bill our users and is easier to reason about.
+    # When we finally figure out which BST quarter we can delete, we convert the time
+    # to UTC so that we delete the correct notifications from our database
+    #
+    # This retention limit is hardcoded and picked from
+    # https://github.com/alphagov/notifications-aws/blob/main/decisions/2022-12-01-notification-history-retention-period.md
+    # and can be increased in the future when we move 3 months into the next financial year
+    retention_limit_bst = datetime(2023, 4, 1)
+
+    oldest_notification_bst = convert_utc_to_bst(db.session.query(func.min(NotificationHistory.created_at)).one()[0])
+    if oldest_notification_bst >= retention_limit_bst:
+        current_app.logger.info(
+            "No notifications older than retention date of %s so nothing to delete", retention_limit_bst
+        )
+        return
+
+    deletion_date_bst = pick_bst_quarter_to_delete_back_from(oldest_notification_bst)
+
+    # in case there is an error in our logic above, let us just double check that we aren't trying to delete for
+    # anything after our retention limit
+    if deletion_date_bst > retention_limit_bst:
+        current_app.logger.exception("Attempted to delete past our notification_history retention limit")
+        raise
```

Comment:
This is really minor (since it does still raise), but calling `raise` outside of a an active Exception like this this gives the error of `RuntimeError: No active exception to reraise` so you could change it to `raise Exception("...")` instead.